pkg/controllers/node/termination/controller.go

```diff
@@ -94,114 +93,273 @@ func (c *Controller) finalize(ctx context.Context, node *corev1.Node) (reconcile
+		return nil, nil
+	}
+	if len(nodeClaims) > 1 {
+		log.FromContext(ctx).WithValues(
+			"node", klog.KObj(node),
+			"nodeclaims", lo.Map(nodeClaims, func(nc *v1.NodeClaim, _ int) klog.ObjectRef { return klog.KObj(nc) }),
+		).Error(fmt.Errorf("duplicate nodeclaims found"), "failed to terminate node")
+		c.recorder.Publish(terminatorevents.DuplicateNodeClaimsFound(node, nodeClaims...))
+		return nil, nil
+	}
+	return nodeClaims[0], nil
+}
+
+type terminationFunc func(context.Context, *v1.NodeClaim, *corev1.Node, *time.Time) (*reconcile.Result, error)
+
+// awaitDrain initiates the drain of the node and will continue to requeue until the node has been drained. If the
+// nodeClaim has a terminationGracePeriod set, pods will be deleted to ensure this function does not requeue past the
+// nodeTerminationTime.
+func (c *Controller) awaitDrain(
+	ctx context.Context,
+	nodeClaim *v1.NodeClaim,
+	node *corev1.Node,
+	nodeTerminationTime *time.Time,
+) (*reconcile.Result, error) {
+	if err := c.terminator.Drain(ctx, node, nodeTerminationTime); err != nil {
 		if !terminator.IsNodeDrainError(err) {
-			return reconcile.Result{}, fmt.Errorf("draining node, %w", err)
+			return &reconcile.Result{}, fmt.Errorf("draining node, %w", err)
 		}
 		c.recorder.Publish(terminatorevents.NodeFailedToDrain(node, err))
-		// If the underlying NodeClaim no longer exists, we want to delete to avoid trying to gracefully draining
-		// on nodes that are no longer alive. We do a check on the Ready condition of the node since, even
-		// though the CloudProvider says the instance is not around, we know that the kubelet process is still running
-		// if the Node Ready condition is true
-		// Similar logic to: https://github.com/kubernetes/kubernetes/blob/3a75a8c8d9e6a1ebd98d8572132e675d4980f184/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L144
-		if nodeutils.GetCondition(node, corev1.NodeReady).Status != corev1.ConditionTrue {
-			if _, err = c.cloudProvider.Get(ctx, node.Spec.ProviderID); err != nil {
-				if cloudprovider.IsNodeClaimNotFoundError(err) {
-					return reconcile.Result{}, c.removeFinalizer(ctx, node)
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetUnknownWithReason(v1.ConditionTypeDrained, "Draining", "Draining"); modified {
+			if err := c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+				if errors.IsConflict(err) {
+					return &reconcile.Result{Requeue: true}, nil
 				}
-				return reconcile.Result{}, fmt.Errorf("getting nodeclaim, %w", err)
+				if errors.IsNotFound(err) {
+					log.FromContext(ctx).WithValues("node", klog.KObj(node)).Error(fmt.Errorf("nodeclaim not found"), "failed to terminate node")
+					c.recorder.Publish(terminatorevents.NodeClaimNotFound(node))
+					return &reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
```

Comment:
The more I think about the amount of error handling you are having to do now in this controller to handle NotFound and do this custom requeue logic for eventing, the more I think a separate controller might be the better place to do this health check to avoid having to handle this everywhere in termination -- also, it would be much more strongly encapsulated than it is right now

Reply:
Discussed offline - we're going to take a middleground approach and not handle the event for now. We'll take a backlog task to consider a separate health controller.